Update ports page (#2126)
gitpod-io/website: gitpod/docs/config-ports.md

# Configure Ports

Gitpod supports exposing HTTP ports via a custom domain that is associated with your workspace. You can also use port forwarding, so that you do not need to update your application if it already references the localhost hostname. You can forward all ports using the [local companion](/docs/ides-and-editors/local-companion), natively in both [VS Code Desktop](/docs/ides-and-editors/vscode), [JetBrains](/docs/ides-and-editors/jetbrains-gateway) and also via the [command-line](/docs/ides-and-editors/command-line) using SSH.

## Default port behaviors

By default, when a port is opening in a Gitpod workspace, Gitpod will:

> 1. **Direct HTTP traffic** - Gitpod exposes newly detected HTTP ports on a private URL. By default, exposed ports require authentication and are available only to those with workspace access. Exposed ports are available on the network of the Gitpod installation (for gitpod.io that's the public internet), which means that ports are accessible from outside the workspace container.

Reviewer: This sentence seemed a little confusing to me.

_**Suggestion (non-blocking)**_: Omit the part about Gitpod network / public internet (or move it to another paragraph), and simplify the text to something like this:

```suggestion
1. **Direct HTTP traffic** - When an application starts listening to an HTTP port, Gitpod detects it and exposes this port on a private URL, which requires authentication. So, by default, this port is available only to those with workspace access. Setting public access to the port, though, makes the URL accessible by anyone.
```

Reviewer 2: Maybe we can also change the wording around "private URL"? I don't think a URL can be private or public. Maybe we can use a word like "guarded", "secured", or "internal"?

Author: > Omit the part about Gitpod network / public internet - @felladrin 

For context: The challenge is that self-hosted installations (e.g. air-gapped) are not necessarily on the public internet, so we need to somehow work that (not assuming public internet) also into the docs (in the least confusing way). 

> I don't think a URL can be private or public - @filiptronicek 

Agree that it's not an industry term, and that might cause confusion. 

However, do consider also that "public/private" is the language we're using in the product, etc to describe a URL that's accessible on a network, or private to that user's workspace.

Author: How about...

```suggestion
1. **Direct HTTP traffic** - When an application starts listening to an HTTP port, Gitpod detects the port and exposes it on a URL that requires authentication. Setting the port to "public" would make the port URL accessible to anyone on the internet (or the installed network if using self-hosted Gitpod).
```

1. **Notify the user** - Gitpod sends the user a popup notification in their [IDE or editor](/docs/ides-and-editors) to let the user know that a port has been detected.

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Setting a port public/private in VS Code Browser" src="/images/editors/port-notification-vscode.png">
    <figcaption>Setting a port public/private in VS Code Browser</figcaption>
</figure>

> <!-- TODO: Show JetBrains notification here (#9908) -->

Author: Now is possible - https://github.com/gitpod-io/gitpod/issues/9908 to be updated

## Accessing port URLs

You can access the dedicated port URL by pre-pending the port number to the workspace URL.

> e.g `3000-yourworkspace.ws-eu45.gitpod.io`

Reviewer 2: Maybe add a dot and capitalize the first letter since the sentence ends before it?

```suggestion
E.g. `3000-yourworkspace.ws-eu45.gitpod.io`
```

Author: Interesting, never knew that — thanks!

You can also print the port URL using the [gp](/docs/command-line-interface) url command e.g. `gp url 3000`.

## Configuring port behaviors

To modify or change default port behaviors, update the `ports` section of your [`.gitpod.yml`](/docs/references/gitpod-yml).

> All changes to ports behaviors take effect immediately, not requiring a workspace restart.

Reviewer 2: I think this might be a typo, but maybe my grammar senses are just bad 😄

```suggestion
All changes to ports' behaviors take effect immediately, not requiring a workspace restart.
```

Author: Yeah, I agree. I think singular "port" probably makes most sense?

```suggestion
All changes to port behaviors take effect immediately, not requiring a workspace restart.
```

**Note:** Some actions (e.g. setting a port public/private) can be taken via the IDE or editor.

### Configure port opening

The port open event is triggered when a new port is detected as open within the workspace.

Port opening behavior can only be set via the [`.gitpod.yml`](/docs/references/gitpod-yml)

The property `onOpen` configures port opening behaviors:

- `notify` (default) - Show a notification for newly detected ports.
- `open-preview` - Open the port URL in a preview within the editor or IDE.
- `open-browser` - Open the port URL a browser tab.

> - `ignore` - Ignore default behavior (notify)

Reviewer 2: Just to make it aligned with the other bullet points

```suggestion
- `ignore` - Ignore default behavior (notify).
```

The property `onOpen` configures port opening behaviors:

- `notify` (default) - Show a notification for newly detected ports.
- `open-preview` - Open the port URL in a preview within the editor or IDE.
- `open-browser` - Open the port URL a browser tab.
- `ignore` - Ignore default behavior (notify)

**Example**: Open a browser tab for port 8080

```yaml
ports:
  - port: 8080
    onOpen: open-browser
```

### Configure port visibility

Port visibility can be set via the [`.gitpod.yml`](/docs/references/gitpod-yml), or manually changed within the IDE or editor.

The property `visibility` configures who can access a port:

- `private` (default) - Only allow users with workspace access to access the port.
- `public` - Allows everyone with the port URL to access the port.

#### Port Visibility in VS Code Browser

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Setting a port public/private in VS Code Browser" src="/images/editors/toggle-port-visibility-vscode.png">
    <figcaption>Setting a port public/private in VS Code Browser</figcaption>
</figure>

#### Port Visibility in VS Code Desktop

> Currently, toggling port visibility is not possible in VS Code Desktop.

Author: This is correct, right @jeanp413 / @akosyakov ? What limitations do we have on the ports view from the SSH Remote Extensions? Can we update it in any way? Or would the best suggestion be for the ports command? 🤔

Reviewer 3: Right now we don't have any control over port forwarding UI in VS Code Desktop. I seen that their port view has visibility by now, but I don't have an idea what it does.

> # Configure Ports

Reviewer 3: Maybe we should have introduction here, i.e.

- remind users about networking in development and production, i.e. usually we will work first on localhost but then deploy on another host, like some public URl on Internet.
- say that Gitpod supports both modes, for compatibility with their existing setup they could use Local port forwarding, or they can adjust their setup to bring it closer to production setup at development time

I'm not sure that many users realize what is Local port forwarding. They may start reconfiguring a project with gp url and so on while they could rather just start Local companion and get ready.

Author: Added: 6b025bae682841718f2a912d91025496d60528ae, let me know what you think !

## Default port behaviors

By default, when a port is opening in a Gitpod workspace, Gitpod will:

1. **Notify the user** - Gitpod sends the user a popup notification in their [IDE or editor](/docs/ides-and-editors) to let the user know that a port has been detected.

> 2. **Direct HTTP traffic** - Gitpod exposes newly detected HTTP ports on a private URL which is available only to those with workspace access.

Author: Added in 6b025bae682841718f2a912d91025496d60528ae, it's a bit hard to word, because if self-hosted it's not necessarily on the public internet :D 

Added...
> Exposed ports are available on the network of the Gitpod installation (for gitpod.io that’s the public internet), which means that ports are accessible from outside the workspace container.

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Setting a port public/private in VS Code Browser" src="/images/editors/port-notification-vscode.png">
    <figcaption>Setting a port public/private in VS Code Browser</figcaption>
</figure>

<!-- TODO: Show JetBrains notification here (#9908) -->

## Accessing port URLs

You can access the dedicated port URL by pre-pending the port number to the workspace URL.

e.g `3000-yourworkspace.ws-eu45.gitpod.io`

You can also print the port URL using the [gp](/docs/command-line-interface) url command e.g. `gp url 3000`.

> If you are using [VS Code in the browser](/docs/ides-and-editors/vscode-browser), any terminal output referencing `localhost:3000` will automatically redirect to the workspace URL.

Reviewer 3: VS Code works the same everywhere:
- in VS Code Browser without local app -> to exposed URL
- in VS Code Browser with local app -> to forwarded URL
- in VS Code Desktop with remote ssh -> to forwarded URL

Author: Will remove for now: c2e2aa86edf0b061e7227743c34a36a39443c2f5

Was kinda [hoping we could link to a VS Code doc](https://github.com/gitpod-io/website/pull/2126/files#r882973585) etc, but will dig around for one in future.

Not essential for the PR.

> <!-- TODO: Link to documentation -->

Author: @jeanp413 - Do you know where this behavior is documented in VS Code? Or the best place to link to?

## Configuring port behaviors

> To modify or change default port behaviors, update the `ports` section of your [`.gitpod.yml`](/docs/references/gitpod-yml).

Author: Added: 1fee60a14a0a9a47b7aa889b9b92b42384230b2f

**Note:** Some actions (e.g. setting a port public/private) can be taken via the IDE or editor.

> ### Configure port opening

Author: Fixed: 84dc4bb040fc15d36bba90f646b80bc85c86abbf

Port opening behavior can only be set via the [`.gitpod.yml`](/docs/references/gitpod-yml)

The property `onOpen` configures port opening behaviors:

- `notify` (default) - Show a notification for newly detected ports.
- `open-preview` - Open the port URL in a preview within the editor or IDE.
- `open-browser` - Open the port URL a browser tab.
- `ignore` - Ignore default behavior (notify)

**Example**: Open a browser tab for port 8080

```yaml
ports:

>   - port: 8080

Reviewer 3: Should we cover port-ranges? name? description?

Author: Let's leave to the following/follow up PR (for simplicity)?

    onOpen: open-browser
```

### Configure port visibility

Port visibility can be set via the [`.gitpod.yml`](/docs/references/gitpod-yml), or manually changed within the IDE or editor.

The property `visibility` configures who can access a port:

- `private` (default) - Only allow users with workspace access to access the port.
- `public` - Allows everyone with the port URL to access the port.

#### Port Visibility in VS Code Browser

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Setting a port public/private in VS Code Browser" src="/images/editors/toggle-port-visibility-vscode.png">
    <figcaption>Setting a port public/private in VS Code Browser</figcaption>
</figure>

#### Port Visibility in VS Code Desktop

Currently, toggling port visibility is not possible in VS Code Desktop.

#### Port Visibility in JetBrains

Currently, toggling port visibility is not possible in JetBrains IDEs.

### Configure port ranges

All port configurations can be applied to ranges as well as single ports.

**Example:** Prevent notifications for ports between 3000 and 8999

```yaml
ports:
  - port: 3000-8999
    onOpen: ignore
```

> ## Port Forwarding

Reviewer 3: `Local Port Forwarding`?

definition below does not cover remote

Author: Updated: 4ba9f889eadb70d7ada7c240aeab1a9e9ac48422

Port forwarding gives you access to ports running within your workspace. Port forwarding is particularly useful when you want to access a remote port using a localhost address.

> ### Port Forwarding in VS Code Desktop

Reviewer 3: `Port Forwarding in VS Code Browser` ?

 it is done with Local Companion

Author: Good point, added! 4ba9f889eadb70d7ada7c240aeab1a9e9ac48422

> With VS Code Desktop, all ports are automatically forwarded, allowing you to access any forwarded ports on your localhost address. You can also manually forward a port using the ports view.

Reviewer 3: I am not sure it is ture, does it mean if we start something in task terminal and connect via VS Code Desktop then it gets forwarded already?

I think it is forwarded on demand when a user interacts with a localhost link via UI or manually, otherwise it is not @jeanp413 correct me if I'm wrong

Reviewer 4: I tested it today and it seems the ports are forwarded automatically

Author: From codespaces docs:

> By default, Codespaces forwards ports using HTTP

https://docs.github.com/en/codespaces/developing-in-codespaces/forwarding-ports-in-your-codespace#about-forwarded-ports

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Setting a port public/private in VS Code Browser" src="/images/editors/port-forwarding-vscode-desktop.png">
    <figcaption>Setting a port public/private in VS Code Browser</figcaption>
</figure>

> ### Port Forwarding in JetBrains IDEs

Reviewer 3: Should not we cover here launch configurations as primary means?

To forward a port in JetBrains, navigate to the preferences page in the [JetBrains Gateway](/docs/ides-and-editors/jetbrains-gateway) client to select the port and protocol to be forwarded.

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Setting a port public/private in VS Code Browser" src="/images/jetbrains-gateway/port-forward-jetbrains.png">

>     <figcaption>Setting a port public/private in VS Code Browser</figcaption>

Reviewer 5: @loujaybee 
Setting a port public/private in ~VS Code Browser~ JetBrains IDE
😄

Author: Oooh, good catch! Thanks Nancy! 

Fixed in: 73f2d1996570b68c3745b3f171a95121800f3c05

</figure>

### Port Forwarding via SSH

> Using SSH command-line access to your workspace, ports can also be forwarded manually using tools such as the OpenSSH remote login client.

Author: @akosyakov / @iQQBot - Can you review this section? Is there anything else you'd add? Should we mention any specifics about the SSH connection, and what is/isn't possible? 

Should we note that remote port forwarding isn't possible? https://github.com/gitpod-io/gitpod/issues/10248

Reviewer 3: It was resolved: https://github.com/gitpod-io/gitpod/pull/10291

Author: > We should link to SSH copy/paste page.

Fixed: ed866235cb8ba73c0327172d164c017db7b7865f

**Example:** Forwarding port `3000` to `localhost:3000`

`ssh -L 3000:localhost:3000 <workspace-ssh-connection>`

> ### Cross-Origin Resource Sharing (CORS)

Author: I feel like this section should be called something else, it's a bit broader than just CORS, as it's also talking about proxying auth headers in requests 🤔

Reviewer 3: It's also only relevant for expose ports, not forwarded. 

> it's also talking about proxying auth headers in requests

I could not find it. It seems to talk about usual CORS configuration for server and client.

Author: Going to restore this section and tackle another day: cceb3956e9d897e9958774f2a3b84c7d968d68f1

By default, when a port is opening in a Gitpod workspace, Gitpod will:

1. **Notify the user** - Gitpod sends the user a popup notification in their [IDE or editor](/docs/ides-and-editors) to let the user know that a port has been detected.
2. **Direct HTTP traffic** - Gitpod exposes newly detected HTTP ports on a private URL which is available only to those with workspace access.

<figure>
<img class="shadow-medium w-full rounded-xl max-w-3xl mt-x-small" alt="Setting a port public/private in VS Code Browser" src="/images/editors/port-notification-vscode.png">
    <figcaption>Setting a port public/private in VS Code Browser</figcaption>
</figure>

<!-- TODO: Show JetBrains notification here -->

## Accessing port URLs

You can access the dedicated port URL by pre-pending the port number to the workspace URL.

e.g `3000-yourworkspace.ws-eu45.gitpod.io`

You can also print the port URL using the [gp](/docs/command-line-interface) url command e.g. `gp url 3000`.

If you are using [VS Code in the browser](/docs/ides-and-editors/vscode-browser), any terminal output referencing `localhost:3000` will automatically redirect to the workspace URL.

<!-- TODO: Link to documentation -->

> <!-- TODO: GIF of behaviour -->

Author: @jeanp413 - Do you know where this behaviour is documented in VS Code?
